go-controller/pkg/util/net_linux.go

```diff
@@ -318,3 +319,28 @@ func DeleteConntrack(ip string, port int32, protocol kapi.Protocol) error {
 	}
 	return nil
 }
+
+// GetNetworkInterfaceIPs returns the IP addresses for the network interface 'iface'.
+func GetNetworkInterfaceIPs(iface string) ([]*net.IPNet, error) {
+	link, err := netLinkOps.LinkByName(iface)
+	if err != nil {
+		return nil, fmt.Errorf("failed to lookup link %s: %v", iface, err)
+	}
+
+	addrs, err := netLinkOps.AddrList(link, netlink.FAMILY_ALL)
+	if err != nil {
+		return nil, fmt.Errorf("failed to list addresses for %q: %v", iface, err)
+	}
+
+	var ips []*net.IPNet
+	for _, addr := range addrs {
+		if addr.Scope != unix.RT_SCOPE_UNIVERSE {
+			continue
+		}
+		if (addr.Flags & unix.IFA_F_DEPRECATED) != 0 {
```

Comment:
maybe a comment on this as to why its here would be helpful in the future. Do we actually have "deprecated" addresses on interfaces in downstream. I've never seen that before.

Reply:
On bare metal jobs where there is no cloud load balancer, it is common to add additional IPs to the node to accept load balancer connections, and to sometimes rebalance those LB IPs between nodes (especially if a node goes down), using ARPs or BGP or whatever to track their movement between nodes.

In the IPv4 case they use `IFA_F_SECONDARY` on these IPs, which seems to automatically cause the IP to move to the end of the list of the interface's IPs, so any code that just picks the first IP will never see those IPs anyway so it was never a problem (though, in the latest version of this patch I added a check for that too, just for completeness).

In IPv6 they use `IFA_F_DEPRECATED` and not `IFA_F_SECONDARY`. (It's possible "secondary" doesn't exist for IPv6 addresses?) "Deprecated" was originally intended for use when you're rolling over IP addresses; you add a new IP, mark the old one as "deprecated", and then when all the connections to the old IP have been closed, and you don't expect any more to come in, you remove it. Among other effects, marking an address as "deprecated" means the kernel will never use it as the source IP of a new outbound connection. So, these semantics are not quite right for a transient load-balancer IP, but they're pretty good; it signals an intention that the IP isn't expected to be around forever, and it keeps the kernel from using that IP for things it wasn't meant for. And the "not expected to be around forever" means we shouldn't use it as the primary node IP either, so it's right to check that flag here.

(Added a comment mentioning bare metal load balancing... people can `git blame` and find the PR for the longer story...)